File: 1605-minimum-number-of-days-to-make-m-bouquets/1605-minimum-number-of-days-to-make-m-bouquets.cpp
class Solution {
public:
    bool possible(vector<int>& bloomDay , int num ,int m,int k){
        int cnt = 0 ;
        int ans = 0 ;
        for(int i = 0 ; i < bloomDay.size() ; i++){
            if(bloomDay[i]<=num)
                cnt++;
            else{
                ans += cnt/k;
                cnt = 0 ;
            }
        }
        ans += cnt/k;       
        return (ans>=m) ;
    }
    int minDays(vector<int>& bloomDay, int m, int k) {
        if(m > bloomDay.size()/k)   return -1 ;
        int ans = -1;
        
        int low = *min_element(bloomDay.begin(),bloomDay.end());
        int high = *max_element(bloomDay.begin(),bloomDay.end());
        
        while(low <= high){
            int mid = low + (high - low)/2;

            if(possible(bloomDay,mid,m,k)==1){
                ans =  mid;
                high = mid - 1 ; 
            }
            else
                low = mid + 1 ;

        }



        return ans;
    }
};

File: 1605-minimum-number-of-days-to-make-m-bouquets/1605-minimum-number-of-days-to-make-m-bouquets.java
class Solution {
    public int minDays(int[] bloomDay, int m, int k) {
        int low = 1 ;
        int high = 0 ;
        int ans = -1 ;
        int total = m * k ;
        if(bloomDay.length < total) return -1 ;
        for(int day : bloomDay){
            high = Math.max(high,day);
        }
        if(bloomDay.length == total) return high ;

    while(low <= high){
        int mid = low + (high - low)/2 ;
        int sum = 0 ;
        int cnt = 0 ;
        for(int i = 0 ; i < bloomDay.length ; i++){
            if(bloomDay[i] <= mid)
                cnt++ ;
            else{
                sum += cnt / k ;
                cnt = 0 ;
            }
        }
        sum += cnt / k ;
        if(sum >= m ){
            ans = mid ;
            high = mid - 1;
        }
        else
            low = mid + 1 ;
    }
    return ans ;
    }
}